### ISSUE TITLE
TBD: Reproduce MSVC/NANT build process (:success:)

### ISSUE THREAD

**Issue by GerHobbelt** [jimmejardine/qiqqa-open-source#14]
🎉 👍 Got an **experimental** Qiqqa binary built using Microsoft Visual Studio 2019 today (2019/august/1).

Also using Syncfusion 17 instead of Syncfusion 14 as that is the release currently available as main choice at Syncfusion for MSVC2019 and I want to use my 'regular' MSVC version instead of having to dual-install or otherwise run an old MSVC setup in parallel on an already pretty loaded dev box.

**happy news**: despite many warnings and errors at the start of this activity a few days ago, I now have a working binary (at least, it works for *me* 😋 🤡 ) and able to modify the source code to hunt down the major showstopping Qiqqa bugs that block me from actually using my PDF collection the way I need!

**Comment by GerHobbelt** [comment]
Using:

- MSVC2019 (the *full* Studio IDE, which I use for other dev work too)
- Syncfusion 17 (instead of version 14)
- Some tweaks to make the reported build errors go away (see upcoming pullreq)
- **WARNING**: the binaries have been copied into the Qiqqa v80 install directory using Beyond Compare in Administrator mode; the `setup.exe` produced by NANT has **NOT BEEN TESTED YET**.

**Comment by GerHobbelt** [comment]
Met a few bumps along the way, but it now works with modern dev tools (MSVS2019).

Turns out NANT doesn't support C#6.0, went all over StackOverflow and elsewhere, did a quick check at https://github.com/nant/nant -> https://github.com/GerHobbelt/nant/commits/master but since that didn't work out within a few hours, I killed that sidetrack and did a hack instead: see [How to build `setup.exe`](https://github.com/GerHobbelt/qiqqa-open-source/blob/a4efd50a3406f5cc4246cfa730e624b67fdfddae/README.md#visual-studio-2019-updated-experimental-build-environment)

First successful result: https://github.com/GerHobbelt/qiqqa-open-source/releases/tag/v81.0.7158.38371

This ends with https://github.com/jimmejardine/qiqqa-open-source/pull/33

**Comment by GerHobbelt** [comment]
Ah, forgot I had this: the set of relevant commits:

----

Revision: e702b0c7885fdf527d0194dbb7f020788a1d57a0
revert/fix NANT build script to produce a `setup.exe` once again.

Revision: 44e157524cbf2eb9a2291979e094795612cf09f9
working on using a more developed build versioning approach. Have MSVS produce a unique version for each build, then (FUTURE WORK) add tooling to ensure all files carry the updated version number(s).

Revision: 177a2be0cff4e92a9ae285c61c2377bac1cbf1c4
Also note that we employ the `?.` **null conditional operator**, which is part of C# 6.0 and described here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-
It turns out the NANT build doesn't cope with this but I'm loath to revert to antiquity there, so the NANT build process has become a little hacky as MSVS2019 (IDE) can build (and debug) the Qiqqa binaries without a fuss, so we now use that one to build the binaries and the NANT build script for packaging (creating the `setup.exe`).

Revision: aac705aab084bea081aaa7d4c25b30c6e6dc5dc9
!@#$%^&* NANT doesn't support C#6.0 language constructs. Spent a few hours on NANT (obtained from GitHub) but cannot get it to work right now. Not spending more time on that distraction: patched the NANT build process to use the binaries which are (correctly and without hassle) produced by MSVC2019 IDE (Batch Build all or Rebuild All, Release target). See short notes in NANT build file and bash shell driver script too. QUICK HACK. QUICK HACK. QUICK HACK. Anyone with the time on their hands and the inclination to make NANT work with a *recent* MSVC setup: more power to you. Right now, this is a bleeding edge NANT rebuild PLUS patched build/packaging scripts.

Revision: 542fc818fb7ac9697e19eea719ce500e6aa2edcb
fix msbuild error (oddly enough this was okay in MSVC2019 and compiled fine in the ID, but the NANT task via `./build-installer.sh` fails.  :-S  )
